luminoso_api/cli.py

```diff
@@ -0,0 +1,123 @@
+import argparse
+import json
+import os
+import sys
+from signal import signal, SIGPIPE, SIG_DFL
+from urllib.parse import urlparse
+
+from .client import LuminosoClient
+from .constants import URL_BASE
+from .errors import LuminosoError
+
+# python raises IOError when reading process (such as `head`) closes a pipe.
+# setting SIG_DFL as the SIGPIPE handler restores behaviour to UNIX default.
+signal(SIGPIPE, SIG_DFL)
+
+
+DESCRIPTION = "Access the luminoso api via the command line."
+
+USAGE = """
+Supply an http verb and a path, with optional parameters.
+Output is returned as json, csv, or an error message.
+
+Parameters may be specified in one of three ways:
+
+A user-friendly key=value parameter list:
+    -p 'key="value"' -p 'key2="value"'
+
+A json object from the command line:
+    -j '{"key": "value", "key2": "value"}'
+
+A file containing a json object:
+    filename.json
+
+Parameter options may be combined. A json object on the command line is merged
+over one in a file, and -p options are merged over both.
+
+GET and DELETE requests append the parameters to the URL.
+POST, PUT, and PATCH send the given parameters as the body of the
+request with Content-Type set to 'application/json'.
+"""
+
+
+def _print_csv(result):
+    """ print a json list of json objects in csv format """
+    first_line = result[0]
+    keys = sorted(list(first_line.keys()))
+    print(','.join(keys))
+    for line in result:
+        print(','.join([str(line[key]) for key in keys]))
+
+
+def _read_params(input_file, json_body, p_params):
+    """ read parameters from input file, -j, and -p arguments, in that order """
+    params = {}
+    try:
+        if input_file:
+            with open(input_file) as f:
+                params.update(json.load(f))
+        if json_body is not None:
+                params.update(json.loads(json_body))
+    except ValueError as e:
+        raise ValueError("input is not valid json: %s" % e)
+    try:
+        params.update({p.split('=', 1)[0]: p.split('=', 1)[1] for p in p_params})
+    except IndexError:
+        raise ValueError("--param arguments must have key=value format")
+    return params
+
+
+def _main():
+    parser = argparse.ArgumentParser(
+        description=DESCRIPTION, epilog=USAGE,
+        formatter_class=argparse.RawDescriptionHelpFormatter)
+    parser.add_argument('-b', '--base-url', default=URL_BASE,
+                        help="api root url, default: %s" % URL_BASE)
+    parser.add_argument('-t', '--token', help="api authentication token")
+    parser.add_argument('-s', '--save-token', action='store_true',
```

Comment:
I stumbled a bit while trying this out because I assumed I wouldn't need to give a method and path if I was just saving a token, but in fact I got an error about missing required arguments:
```
$ lumi-api -s -t 'my-token'
usage: lumi-api [-h] [-b BASE_URL] [-t TOKEN] [-s] [-p PARAM] [-j JSON_BODY]
                [-c]
                {get,post,put,patch,delete} path [input_file]
lumi-api: error: the following arguments are required: method, path
```

Is there a workable way to make those arguments optional when saving the token, short of throwing out argparse altogether?